exe-unit/src/network/inet.rs

```diff
@@ -427,10 +495,47 @@ impl Proxy {
 
     async fn get(&self, key: &TransportKey) -> Option<TransportSender> {
         let state = self.state.read().await;
-        state.remotes.get(key).cloned()
+        state.remotes.get(key).map(|state| state.sender.clone())
+    }
+
+    async fn update_seen(&self, key: &TransportKey) -> bool {
+        let state = self.state.read().await;
+        state
+            .remotes
+            .get(key)
+            .map(|state| state.update_seen())
+            .is_some()
     }
 
-    async fn bind(&self, desc: SocketDesc) -> Result<impl Future<Output = ()> + 'static> {
+    async fn drop_unused_connections(&self, sockets_limit: usize) -> usize {
+        let udps = {
+            let state = self.state.read().await;
+            let mut udps = state
+                .remotes
+                .iter()
+                .filter(|(key, _conn)| key.0 == Some(Protocol::Udp))
+                .collect::<Vec<_>>();
+            let to_remove = udps.len().saturating_sub(sockets_limit);
+
+            udps.sort_by_key(|element| element.1.last_seen.load(Ordering::Relaxed));
+
+            udps[0..to_remove]
+                .iter()
+                .map(|(key, _)| (*key).clone())
+                .collect::<Vec<_>>()
+        };
+
+        for key in &udps {
+            self.unbind(key.as_socket_desc()).await.ok();
```

Comment:
> Maybe there could be an unbind_many function which drains the state in one lock and closes/aborts connections in parallel?

I don't think we need that.
1. udps.len() will always be 1 with the current usage, so we don't get any performance gain from this.
2. `Proxy::unbind` is relatively cheap to call

All in all I don't think the additional complexity here is justified.

> I am not sure if inet_ingress_handler can be called on another thread, but if yes, then there seems to be a race condition where this loop will close a socket that was active in the meantime.

Not sure how to solve it, even with atomic `unbind_many` we would still need to block `inet_ingress_handler` from proceeding until we are done removing sockets. Don't know if this can actually introduce a bug, will need to reconsider.

Reply:
I've had a talk with witek regarding this issue, and we've concluded that this inaccuracy doesn't arise until we start processing requests concurrently. Added a comment noting this dependency in the code.

Reply:
Added a comment in Router::handle